Optionals v3 split and simplify

joshring opened the pull request:
- Drastically simplifying optionals and splitting into a basic section and an advanced section
- Improving the language to be clearer, `Empty` Optionals and `Result` found etc

joshring committed:
hotfix of indexing

joshring committed:
Revert "hotfix of indexing"

This reverts commit 352b5947bbf74a88bc62efc942c867d6b8e7f6cc.

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Merge remote-tracking branch 'origin/main'

joshring committed:
Drastically simplifying optionals and splitting into a basic section and an advanced section

improving the language to be clearer

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+    order: 115
+---
+
+These features are *nice to haves*, you should be able to do what you need mostly using the [standard Optionals features](./optionals_intro.md).
```
The link here looks wrong?

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+fn char[]! read_file(String filename, char[] buffer)
+{
+    // Return `Excuse` if failed opening file, using rethrow `!`
+    File! file = file::open(filename, "r")!; 
```
This should not be `File!`, just `File` if you did the rethrow.

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+    if (catch excuse = read_buffer) 
+    {
+        io::printfn("Excuse found: %s", excuse);
+        return excuse?; // Returning `Excuse` using `?` suffix
```
This comment looks long, place it above if this happens, so there is no need to scroll the code frame on smaller screens.

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+
+### Using force unwrap `!!` to panic on `Empty` Optional
+
+The force unwrap `!!` will issue a panic and exit the program if Optional is `Empty`.
```
Add something like "this is useful when the error should – in normal cases – not happen and you don't want to write any error handling for it. That said, it should be used with caution in production code"

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+
+// Find `Empty` Optional handling inside scope
+if (catch optional_value) 
+{
```
When the `if` isn't in a real function, the syntax highlighting might have issues, so best practice is to put it in a proper function which can be as simple as `fn void test() { ... }`

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
````diff
@@ -0,0 +1,355 @@
+}
+```
+
+## Find `Result` is present 
````
"Check if a `Result` is present" or something like that.

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
````diff
@@ -0,0 +1,355 @@
+}  
+```
+
+For example
````
Do you mean "For example:"?

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
````diff
@@ -0,0 +1,355 @@
+}
+```
+
+### Assert an Optional cannot have a missing `Result` and the program should panic with force unwrap `!!`
````
This title is hard to read.

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+The `void!` type has no possible representation as a variable, and may
+only be a function return type. 
+
+Use `if (catch)` to handle an `Empty` Optional or read the `Excuse`
```
This seems to repeat the text higher up.

joshring replied:
Yes, I was thinking it could be helpful to give working examples for void! returns as they can be a bit confusing, will remove

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
````diff
@@ -0,0 +1,355 @@
+```
+
+
+## Default assign when Optional is `Empty` using `??`
````
This part should be higher up, as it's important to learn of this operator. Also, the title needs to be improved, "Return a default value when `Empty`" or something.

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
````diff
@@ -0,0 +1,355 @@
+int regular_value = foo_may_error() ?? -1;
+```
+
+### Signpost the unique origin of the same `Excuse`
````
Signpost?

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -0,0 +1,355 @@
+
+    // We can tell these appart by default assigning our own unique `Excuse`
+    // Our unique `Excuse` is assigned only if a `Excuse` is found
+    int c = test() ?? NoHomework.DOG_ATE_MY_HOMEWORK?;
```
This is incorrect, maybe you meant:
`int! c = test() ?? NoHomework.DOG_ATE_MY_HOMEWORK?;`

lerno replied:
And same with "d" below.

joshring committed:
comments

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -19,6 +19,73 @@ fn int! example() { ... }
 ## Handling `Empty` Optional
 
+### Return a default value when Optional is `Empty`
+If an expression returns a `Empty` Optional default assign using `??` operator.
```
"The `??` operator allows us to return a default value if the result if the optional is `Empty`.

lerno commented on src/content/docs/references/docs/optionals_advanced.md:
```diff
@@ -19,6 +19,73 @@ fn int! example() { ... }
+
+#### Make origin of common `Excuse` explicit
+
+Catch a `Empty` Optional and change the `Excuse` to a more specific `Excuse`, that allows us to disitinguish one situation from the other, even with the same starting `Excuse`.
```
"an `Empty`"

joshring replied:
Found some more occurrences will fix

joshring committed:
grammar
improved ?? description

joshring committed:
smaller grammar and formatting changes

joshring committed:
removing unneeded sentence

joshring committed:
adding link back to basic optionals page from @catch

joshring committed:
adding link to basic optionals page
putting example higher on the page

joshring committed:
renaming instance of "result" -> "Optional"

joshring committed:
broken link

joshring committed:
named arguments syntax change

joshring committed:
fix favicon in docs

joshring committed:
fix favicon

joshring committed:
updating docs to use new named arguments

joshring committed:
fixing optionals link in defer
adding link to rethrow inside of optionals page itself

joshring committed:
Merge branch 'c3lang:main' into optionals_v3_split_and_simplify

lerno committed:
Formatting and naming changes. Correction on '!' meaning / behaviour.

joshring committed:
- updating links to be correct
- all terms have consistent quoting like: “Result”
- ex is hardly shorter than excuse so I went with that as it's clearer

joshring committed:
formatting inline with discussion
